docs/source/development/AlgorithmMPISupport.rst

```diff
@@ -0,0 +1,464 @@
+The default implementation of ``Algorithm::execNonMaster()`` creates an **uninitialized** (in the case of ``MatrixWorkspace``) workspace of the same type as the input workspace.
+If ``Algorithm::execNonMaster()`` is overridden, any workspaces that are created shall also be uninitialized and should have storage mode ``StorageMode::MasterOnly``.
+
+Given that the workspace on non-master ranks are not initialized, no methods of the workspace should be called, apart from ``Workspace::storageMode()``.
+Validators on the non-master ranks are thus also disabled.
+
+A typical implementation could look as follows:
+
+.. code-block:: c++
+
+  void MyAlg::execNonMaster() {
+    setProperty("OutputWorkspace", Kernel::make_unique<Workspace2D>(
+                                       Parallel::StorageMode::MasterOnly));
+  }
+
+
+Setting spectrum numbers
+------------------------
+
+Setting spectrum numbers via the legacy interface ``MatrixWorkspace::getSpectrum(size_t)::setSpectrumNo(specnum_t)`` is not supported in MPI runs and will throw an exception.
+The reason is that spectrum numbers are used to globally identify a spectrum and thus changing a spectrum number must be done globally, i.e., on all MPI ranks.
+Spectrum numbers should be set by using ``Indexing::IndexInfo`` and ``MatrixWorkspace::setIndexInfo()``, or rather by passing the ``IndexInfo`` to one of the workspace factory functions from ``DataObjects/WorkspaceCreation.h``.
+
+
+Workspace indices
+-----------------
+
+If a workspace is distributed, i.e., has storage mode ``StorageMode::Distributed`` workspaces indices lose their meaning.
+In particular, ``MatrixWorkspace::getNumberHistograms()`` will return the local number of spectra and not the global size of the workspace.
+For purposes of interaction with the user interface and for internal consistency a global equivalent of the 'workspace index' concept has been introduced.
+This index is represented by ``Indexing::GlobalSpectrumIndex``. [#spectrum-index]_
+
+The consequences are as follows:
+
+- Workspace indices should not be logged or written into output of other types such as tables. Instead spectrum numbers (``Indexing::SpectrumNumber``) or global spectrum indices (``Indexing::GlobalSpectrumIndex``) must be used.
+- The number of histograms in a workspace obtained from ``MatrixWorkspace::getNumberHistograms()`` may only be used for processing all spectra.
+  It should not be logged, written as output, or used for branching execution paths since it is meaningless.
+  If the total number of spectra in a workspace is required it can be accessed via ``MatrixWorkspace::indexInfo()::globalSize()``.
+- User input providing indices or spectrum numbers in way or another must be translated into local indices by ``IndexInfo``.
+  The most common cases will be covered by a workspace property that also accepts indices (under development).
```

Comment:
@LamarMoore you should create a placeholder in your work to ensure that this gets updated once you have completed the new property.